### PR TITLE
[TT-4793] request and res log obfuscation

### PR DESCRIPTION
**Pull request by urbanishimwe**
This is an updated version of https://github.com/TykTechnologies/tyk/pull/3896
It contains most of the changes made by @joshblakeley and some refactoring.

## Description
Enables editing or removal of all parts of analytics records and raw request and responses recorded by Tyk at the gateway level.

## Related Issue
https://tyktech.atlassian.net/browse/TT-4793

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit:  
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2071786371)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: 36c8ee7819c060e0ae5803beaa739e5ffd02a673 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2090394820)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: e73b1ccc1a9c993378a7a11bf3f6af777e1d37ff 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2090527173)

**Review by titpetric — CHANGES_REQUESTED** [review]

File: gateway/analytics.go [hunk 1]
```diff
@@ -336,3 +199,42 @@ func (r *RedisAnalyticsHandler) recordWorker() {
...
+	if ipStr == "" {
+		return nil, nil
+	}
+	ip := net.ParseIP(ipStr)
+	if ip == nil {
+		return nil, fmt.Errorf("invalid IP address %q", ipStr)
+	}
+	record := new(analytics.GeoData)
+	if err := gw.Analytics.GeoIPDB.Lookup(ip, record); err != nil {
+		return nil, fmt.Errorf("geoIPDB lookup of %q failed: %v", ipStr, err)
```
[the remark this answers is not in the document]
Comment (urbanishimwe) [reply gateway/analytics.go:237]: Another Linter report this as wrong format arg for type interface{}

File: gateway/analytics_go_plugin.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	return true
+}
+
+func (m *GoAnalyticsPlugin) processRecord(record *analytics.Record) (err error) {
+	// make sure tyk recover in case Go-plugin function panics
+	defer func() {
+		if e := recover(); e != nil {
+			err = fmt.Errorf("%v", e)
```
[the remark this answers is not in the document]
Comment (urbanishimwe) [reply gateway/analytics_go_plugin.go:43]: Another Linter report this as wrong format arg for type interface{}

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: 6675cf3b1158cedf440aa9ff205bf52677feafa3 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2095873781)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: 2dadfc0d4d4fd4b3a2e6d1a7ff6d9e9cbff8197a 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2095882567)

**Review by titpetric — APPROVED** [review]

**Review by jeffy-mathew — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: 1fc116e6752c605073d53f213a628e4b92390a44 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2114180827)

File: gateway/analytics.go [hunk 3]
```diff
@@ -336,3 +199,42 @@ func (r *RedisAnalyticsHandler) recordWorker() {
...
+	// Not great, tightly coupled
+	if gw.Analytics.GeoIPDB == nil {
+		return
+	}
+
+	record, err := geoIPLookup(ipStr, gw)
+	if err != nil {
+		log.Error("GeoIP Failure (not recorded): ", err)
+		return
+	}
```
Comment (furkansenharputlu) [inline gateway/analytics.go:213]: Can you please skip line after `if` and `for` blocks, it makes the code unreadable.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: a58bfb2da1c88b4ffd70d084e4684067a06f26ad 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2148073699)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: df31ce6d8123d054c1dc9ec408f8c394392d8518 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2149493049)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: f848ea88a1488fd29bcadd8978970833f825c06a 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2149628243)

**Review by furkansenharputlu — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: 4e480a9963934e097e86ac20f9437be5f507d5ce 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2154612558)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: ef9a52adc61c4ed06f11cdeec5ba99064ab01f82 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2154643590)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: 074cbcc3d8ada653edd72ab0a11fa60fac101218 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2154995764)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: 38e8f620e24e39cdab172789de069f41bca628e6 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2167418625)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/3981/merge
Commit: 0780bdf953863edc9ecc637735694020bef93b0c 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2183696117)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3981/merge
Commit: 0780bdf953863edc9ecc637735694020bef93b0c 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2183696117)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 4a365db93acf41289d726900e44b4609877e7ce9 [TT-4793] request and res log obfuscation (#3981)

This is an updated version of https://github.com/TykTechnologies/tyk/pull/3896
It contains most of the changes made by @joshblakeley and some refactoring.

## Description
Enables editing or removal of all parts of analytics records and raw request and responses recorded by Tyk at the gateway level.

## Related Issue
https://tyktech.atlassian.net/browse/TT-4793

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

Co-authored-by: joshblakeley <josh@tyk.io>
Co-authored-by: Jeff <jeffy.mathew100@gmail.com>
Triggered by: push (@buger)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2183903231)

**Pull request merged**